### DISCUSSION TITLE
`async`/`await` code in `config.js` to load secrets from external source on the cloud

### DISCUSSION THREAD

**Discussion by SonNg2k** [directus/directus#16633] [Feature Requests]
I want to use top-level `await` within the `config.js` file to asynchronously load secrets from external source on the cloud but [Directus' CommonJS JavaScript package generated as a result of TypeScript compilation doesn't allow this](https://github.com/directus/directus/blob/d6b8e5963d955769cfe2a721014ca17191d8850c/api/tsconfig.json#L3). As a result, the following command `CONFIG_PATH="config.js" npx directus start` results in the following error:

```
/node_modules/directus/dist/env.js:298
        const module = require(configPath);
                       ^

Error [ERR_REQUIRE_ESM]: require() of ES Module /config.js from /node_modules/directus/dist/env.js not supported.
config.js is treated as an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which declares all .js files in that package scope as ES modules.
Instead rename config.js to end in .cjs, change the requiring code to use dynamic import() which is available in all CommonJS modules, or change "type": "module" to "type": "commonjs" in /package.json to treat all .js files as CommonJS (using .mjs for all ES modules instead).

    at processConfiguration (/node_modules/directus/dist/env.js:298:24)
    at Object.<anonymous> (/node_modules/directus/dist/env.js:268:8)
    at Object.<anonymous> (/node_modules/directus/dist/cache.js:9:31)
    at Object.<anonymous> (/node_modules/directus/dist/middleware/respond.js:8:17)
    at Object.<anonymous> (/node_modules/directus/dist/controllers/activity.js:10:19)
    at Object.<anonymous> (/node_modules/directus/dist/app.js:35:36)
    at Object.<anonymous> (/node_modules/directus/dist/server.js:36:31)
    at Object.<anonymous> (/node_modules/directus/dist/cli/index.js:8:18)
    at Object.<anonymous> (/node_modules/directus/dist/cli/run.js:3:17)
    at Object.<anonymous> (/node_modules/directus/cli.js:2:1) {
  code: 'ERR_REQUIRE_ESM'
}
```

My `config.js` file is as below:

```javascript
import {SecretManagerServiceClient} from '@google-cloud/secret-manager';

const client = new SecretManagerServiceClient();

const dbEndpoint = await getSecret(
  'projects/cloud_project_id/secrets/directus_aws_rds_endpoint/versions/latest');
const dbPass = await getSecret(
  'projects/cloud_project_id/secrets/directus_aws_rds_password/versions/latest');
const s3Key = await getSecret(
  'projects/cloud_project_id/secrets/directus_aws_s3_key/versions/latest');
const s3Secret = await getSecret(
  'projects/cloud_project_id/secrets/directus_aws_s3_secret/versions/latest');

async function getSecret(secretRef) {
  const [version] = await client.accessSecretVersion({
    name: secretRef
  });
  return version.payload?.data?.toString() || '';
}

export default {
  DB_CLIENT: 'pg',
  DB_HOST: dbEndpoint,
  DB_PORT: 5432,
  DB_DATABASE: 'directus',
  DB_USER: 'postgres',
  DB_PASSWORD: dbPass,
  DB_SSL: false

  // More config options here...
};
```

The `tsc` compilation option of Directus' TypeScript codebase should be changed to support the top-level `await` syntax inside `config.json`. [Top-level `'await'` expressions are only allowed when the `'module'` option is set to `'es2022'`, `'esnext'`, `'system'`, `'node16'`, or `'nodenext'`, and the `'target'` option is set to `'es2017'` or higher](https://stackoverflow.com/questions/63718054/how-to-use-top-level-async-await-with-typescript).

According to [the best practice from Google Cloud](https://cloud.google.com/run/docs/configuring/environment-variables):

> Caution: You should not use environment variables to store and consume secrets, because environment variables are visible to anyone with project Viewer permissions or greater. Instead, use Secret Manager with Cloud Run as described in the [Using secrets](https://cloud.google.com/run/docs/configuring/secrets) page.

💡 Another possible solution: allow `config.js` to export a JavaScript Promise with `module.exports`.

**Comment by paescuj** [comment]
Generally speaking, I'm all in for async config files.
Unfortunately, it's not as simple as adjusting the `tsconfig.json` file. The whole environment / configuration loading implementation would have to be rewritten (to be async as well) and most likely this means that the migration of the API code-base to ESM must be completed first (see #15190).

As a workaround for now you could, for example, run a pre-script which loads the config and stores the secrets in files before starting the Directus instance. Note that Directus supports [loading secrets from files](https://docs.directus.io/self-hosted/config-options.html#environment-variable-files).

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
